src/pg_copy.c

```diff
@@ -0,0 +1,742 @@
+	tupleDescriptor = RelationGetDescr(rel);
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	/*
+	 * We create a new memory context called tuple context, and read and write
+	 * each row's values within this memory context. After each read and write,
+	 * we reset the memory context. That way, we immediately release memory
+	 * allocated for each row, and don't bloat memory usage with large input
+	 * files.
+	 */
+	tupleContext = AllocSetContextCreate(CurrentMemoryContext,
+										 "COPY Row Memory Context",
+										 ALLOCSET_DEFAULT_MINSIZE,
+										 ALLOCSET_DEFAULT_INITSIZE,
+										 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * Construct hash table used for shardId->Connection mapping.
+	 * We can not use connection cache from connection.c used by GteConnection because
+	 * we need to establish multiple connections with each nodes: one connection per shard
+	 */
+	shardToConn = CreateShardToConnectionHash();
+
+	/* init state to read from COPY data source */
+	copyState = BeginCopyFrom(rel, copyStatement->filename,
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+
+	if (copyState->binary)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("Copy in binary mode is not currently supported")));
+	}
+	PG_TRY();
+	{
+		/* Lock all shards in shared mode */
+		shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+
+		shardCount = shardIntervalList->length;
+		shardListCache = palloc0(shardCount * sizeof(List *));
+		hashTokenIncrement = (uint32) (HASH_TOKEN_COUNT / shardCount);
+
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+		}
+
+		while (true)
+		{
+			ShardInterval *shardInterval = NULL;
+			ShardId shardId = 0;
+			int errorCount = 0;
+			bool found = false;
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno-1])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+			partitionColumnValue = columnValues[partitionColumn->varattno-1];
+			hashedValue = DatumGetInt32(FunctionCall1(hashFunction,
+													  partitionColumnValue));
+			shardHashCode =
+				(int) ((uint32) (hashedValue - INT32_MIN) / hashTokenIncrement);
+			prunedList = shardListCache[shardHashCode];
+
+			if (prunedList == NULL)
```

Comment:
There are many partitionColumnValues that map to the same shardHashCode. The first row that maps to a shardHashCode might match the interval and cause the interval to be cached for that shardHashCode, but the second row that maps to the same shardHashCode might not fall within the interval.